### PR TITLE
build: Remove required dependency of Pandas

### LINKED ISSUE #14683: Drop Pandas as a required dependency
### Problem description

With the work done in https://github.com/bokeh/bokeh/pull/14169 and  https://github.com/bokeh/bokeh/pull/14677, it no longer seems necessary to have pandas as a required dependency. 

My plan would therefore be to remove it as a dependency and update the codebase to handle an environment without pandas installed. This will likely be a combination of `sys.modules` checks and a one-time check with `importlib.util.find_spec`, so we don't check `sys.modules` for pandas when we are in an environment without it. 

I'm not sure whether we should add a test run with minimum dependencies and only pytest.  

### Feature description

`-`

### Potential alternatives

`-`

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by hoxbro**
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [x] issues: fixes #14683
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by hoxbro** [comment]
[29ec7a9](https://github.com/bokeh/bokeh/pull/14686/commits/29ec7a9915f27007af268a7fe37e530e08a94942) add a new test case called core, which only installs the required dependencies + pytest. 

I feel there is validity to this kind of test run, but I can remove it if people don't agree.

File: tests/unit/bokeh/io/test_webdriver.py [hunk 1]
```diff
@@ -12,6 +12,9 @@
 
 import pytest ; pytest
 
+if not pytest.importorskip("selenium"):
```
Comment (hoxbro) [inline tests/unit/bokeh/io/test_webdriver.py:15]: Using `pytest.importorskip` here because I don't want to mess up imports below

File: tests/unit/bokeh/models/test_sources.py
Comment (hoxbro) [inline tests/unit/bokeh/models/test_sources.py:1]: The large change in this file is moving pandas-related tests into a new class.

File: .github/workflows/bokeh-ci.yml [hunk 2]
```diff
@@ -223,6 +223,47 @@ jobs:
           flags: unit,minimal
           verbose: true
 
+  core-deps:
```
Comment (hoxbro) [inline .github/workflows/bokeh-ci.yml:226]: Most of this is copied for `minimal-deps`

File: src/bokeh/models/sources.py [hunk 3]
```diff
@@ -552,11 +558,11 @@ def _stream(self, new_data: DataDict | pd.Series | pd.DataFrame,
             source.stream(new_data)
 
         '''
-        import pandas as pd
+        pd = sys.modules.get("pandas")
```
Comment (hoxbro) [inline src/bokeh/models/sources.py:561]: Can't use `uses_pandas` here because we loop through a dict further down.

File: src/bokeh/plotting/_plot.py [hunk 4]
```diff
@@ -78,16 +79,19 @@
...
+    if uses_pandas(range_input):
+        import pandas as pd
+        from pandas.core.groupby import GroupBy
+    else:
+        pd = GroupBy = None
 
     if range_input is None:
         return DataRange1d()
-    if isinstance(range_input, GroupBy):
+    if GroupBy and isinstance(range_input, GroupBy):
```
Comment (hoxbro) [inline src/bokeh/plotting/_plot.py:90]:
```suggestion
    if pd and isinstance(range_input, GroupBy):
```

I think this is better.

File: src/bokeh/util/dependencies.py [hunk 5]
```diff
@@ -102,6 +106,11 @@ def uses_pandas(obj: Any) -> bool:
 # Private API
 #-----------------------------------------------------------------------------
 
+@cache
+def _is_installed(mod_name: str) -> bool:
+    mod_name, *_ = mod_name.split(".")
```
Comment (hoxbro) [inline src/bokeh/util/dependencies.py:107]: The split is to avoid importing the module. 

image

File: src/bokeh/util/serialization.py [hunk 6]
```diff
@@ -52,19 +56,20 @@
...
+        result.add(pd.Timestamp)
+        result.add(pd.Timedelta)
+        result.add(pd.Period)
+        result.add(type(pd.NaT))
     return result
 
 def __getattr__(name: str) -> Any:
     if name == "DATETIME_TYPES":
-        return _compute_datetime_types()
+        return _compute_datetime_types(_is_installed("pandas") and "pandas" in sys.modules)
```
Comment (hoxbro) [inline src/bokeh/util/serialization.py]: We could also just use `sys.modules`.

```suggestion
        return _compute_datetime_types("pandas" in sys.modules)
```

File: tests/unit/bokeh/embed/test_standalone.py [hunk 7]
```diff
@@ -82,6 +83,8 @@ def test_plot() -> figure:
 </body>
 """)
 
+CSS_SELECTOR: By.CSS_SELECTOR = "css selector"  # type: ignore[valid-type]
```
Comment (hoxbro) [inline tests/unit/bokeh/embed/test_standalone.py:86]: A bit dirty, but we have a test to verify that this should always match.

**Comment by hoxbro** [comment]
This should be in a good state to review. I have left some comments in the code; some are where I need an opinion, and others explain why I have done what I have done.

File: src/bokeh/plotting/_plot.py (hunk 4 above)
Comment (mattpap) [reply src/bokeh/plotting/_plot.py:90]: Then `GroupBy` can be `None` (from typing perspective) after `and`.

File: src/bokeh/util/serialization.py [hunk 8]
```diff
@@ -176,22 +181,26 @@ def convert_datetime_type(obj: Any | pd.Timestamp | pd.Timedelta | dt.datetime |
...
 
     # Pandas Timestamp
-    if isinstance(obj, pd.Timestamp):
+    if pd and isinstance(obj, pd.Timestamp):
         return obj.value / 10**6.0
 
     # Pandas Timedelta
-    elif isinstance(obj, pd.Timedelta):
+    elif pd and isinstance(obj, pd.Timedelta):
         return obj.value / 10**6.0
```
Comment (mattpap) [inline src/bokeh/util/serialization.py]: I can't add a suggestion change, so this will have to be applied manually. This seems much simpler and easier on typing:

```py
    if uses_pandas(obj):
        import pandas as pd

        # Pandas NaT
        if obj is pd.NaT:
            return np.nan

        # Pandas Period
        if isinstance(obj, pd.Period):
            return obj.to_timestamp().value / 10**6.0

        # Pandas Timestamp
        if isinstance(obj, pd.Timestamp):
            return obj.value / 10**6.0

        # Pandas Timedelta
        if isinstance(obj, pd.Timedelta):
            return obj.value / 10**6.0
```

Comment (mattpap) [reply src/bokeh/util/serialization.py]: For consistency I would also change `elif` to `if` in the rest of this function.

File: tests/codebase/test_no_tornado_common.py [hunk 9]
```diff
@@ -22,20 +22,29 @@
 from sys import executable as python
 
 # Bokeh imports
+from bokeh.util.dependencies import _is_installed
```
Comment (mattpap) [inline tests/codebase/test_no_tornado_common.py]: I'm getting warnings about this being a private function used outside of its scope. Given this is used across bokeh and its tests, then I would remove `_` prefix and move it to `dev` section (from `private`).

File: src/bokeh/util/serialization.py (hunk 6 above)
Comment (mattpap) [reply src/bokeh/util/serialization.py]: I don't see why this has to be an argument to this function altogether. I would just compute this boolean within this function.

File: src/bokeh/models/util/structure.py [hunk 10]
```diff
@@ -300,22 +298,22 @@ def _make_graph_plot(self) -> Plot:
...
-            Z["types"] = Z["types"].map(lambda x: str(x))
-            df = pd.concat([df, Z])
-        return df
+            Z = self._obj_props_to_df2(M)
+            n_items = len(Z["values"])
+            dct["id"].extend([str(x[0])] * n_items)
+            dct["model"].extend([str(M)] * n_items)
+            dct["values"].extend(list(map(str, Z["values"])))
+            dct["types"].extend(list(map(str, Z["types"])))
+            dct["props"].extend(Z["props"])
```
Comment (mattpap) [inline src/bokeh/models/util/structure.py:315]: This would be a pretty good case for a dataclass, but then I suppose it would be a breaking change.

**Review by mattpap — COMMENTED** [review]
Functionally looks good, with minor issues to resolve.

File: src/bokeh/util/serialization.py (hunk 6 above)
Comment (hoxbro) [reply src/bokeh/util/serialization.py]: `_compute_datetime_types` is decorated with `lru_cache`

File: tests/codebase/test_no_tornado_common.py (hunk 9 above)
Comment (hoxbro) [reply tests/codebase/test_no_tornado_common.py]: I don't think we should remove the `_`, I still don't think any user should use this function. 

Will move it to dev section.

Comment (bryevdv) [reply tests/codebase/test_no_tornado_common.py]: If I were doing this today I would have the underscore on the module `_util` and then all the functions inside don't need them. Generally, functions with underscores should only be used in the module they are defined in.

Comment (hoxbro) [reply tests/codebase/test_no_tornado_common.py]: Renamed in [f4ed93a](https://github.com/bokeh/bokeh/pull/14686/commits/f4ed93a926cf9e8283bac63aebd7b4c1809e2619)

**Review by mattpap — APPROVED** [review]
I will publish dev4 as soon as this is merged.
